SIS_import/sis_upload.rb

```diff
@@ -4,93 +4,151 @@
 require "fileutils"
 require "rubygems"
 require "nokogiri"
+require "digest"
 
 
-def upload_to_canvas(fileName, token, server, outputDirectory)
+def upload_to_canvas(currentDirectory, fileName, token, server, outputDirectory, outputFile, output_file_base_name)
 
-	# get file names
-	currentFileBaseName = File.basename(fileName)
-	currentFileBaseNameWithoutExtension = File.basename(fileName, ".zip")
 
-
-	#set the error flag, default to be false
+	# set the error flag, default to be false
 	upload_error = false
 
-	#open the output file
-	begin
-		outputFile = File.open(outputDirectory + currentFileBaseNameWithoutExtension + ".txt", "w")
+	# prior to upload current zip file, make an attempt to check the prior upload, whether it is finished successfully
+	if (prior_upload_error(currentDirectory, server, token))
+		## check first about the environment variable setting for MAILTO '
+		return "Previous upload job has not finished yet."
+	end
 
-		# Web Service call
-		p fileName
-		p server
+	# upload start time
+	outputFile.write("upload start time : " + Time.new.inspect)
 
-		json_data=`curl -H "Content-Type: application/zip" --data-binary @#{fileName} -H "Authorization: Bearer #{token}" #{server}/api/v1/accounts/1/sis_imports.json?import_type=instructure_csv`
+	# continue the current upload process
 
-		outputFile.write("#{json_data}\n")
-		parsed = parseJson(json_data)
+	# Web Service call
+	p fileName
+	p server
 
-		if (parsed["errors"])
-			## break and print error
-			error_array=parsed["errors"]
-			## hashmap ["message"=>"error_message"
-			uploadError = error_array[0]["message"]
-			outputFile.write("upload error: " + uploadError)
-			outputFile.write("\n")
+	json_data=`curl -H "Content-Type: application/zip" --data-binary @#{fileName} -H "Authorization: Bearer #{token}" #{server}/api/v1/accounts/1/sis_imports.json?import_type=instructure_csv`
```

Comment:
I believe I may have mentioned this before.  Calling curl in this way is not a very good idea.  It adds an external dependency to the application.  This call also starts up another shell which will then run curl, which makes for at least two additional processes, maybe more.  Those processes could die or be terminated, which means nothing will be returned to this program.  However, the value of json_data isn't checked at all.  The call to curl in another shell should be replaced with a call to a native Ruby HTTP client.  If that's done, there will probably no longer be a need to write data to a file to pass to the HTTP client.  It will probably accept the data as part of the function/method call.